src/PAO_grid_transform_module.f90

```diff
@@ -107,135 +107,145 @@ subroutine single_PAO_to_grid(pao_fns)
+    real(double), allocatable, dimension(:) :: x_store, y_store, z_store, r_store ! outputs of check_block
+
+    nblock = domain%groups_on_node
+    npart = maxval(naba_atoms_of_blocks(atomf)%no_of_part)
+    natom = maxval(naba_atoms_of_blocks(atomf)%no_atom_on_part)
 
     call start_timer(tmr_std_basis)
     call start_timer(tmr_std_allocation)
-    allocate(ip_store(n_pts_in_block),x_store(n_pts_in_block),y_store(n_pts_in_block),z_store(n_pts_in_block), &
-         r_store(n_pts_in_block))
+    allocate(ip_store(n_pts_in_block ))
+    allocate(x_store( n_pts_in_block ))
+    allocate(y_store( n_pts_in_block ))
+    allocate(z_store( n_pts_in_block ))
+    allocate(r_store( n_pts_in_block ))
+    allocate(offset_position(natom, npart, nblock))
     call stop_timer(tmr_std_allocation)
     ! --  Start of subroutine  ---
 
+    ! No need to compute these here, since they are derived from globals
+    ! Store with rcellx, rcelly, rcellz?
+    ! Reduces code duplication
     dcellx_block=rcellx/blocks%ngcellx
     dcelly_block=rcelly/blocks%ngcelly
     dcellz_block=rcellz/blocks%ngcellz
 
     call my_barrier()
 
-    no_of_ib_ia = 0
     gridfunctions(pao_fns)%griddata = zero
+    next_offset_position = 0
+    offset_position = huge(0)
+    rcut = r_h + RD_ERR
 
-    max_num_blocks = maxval(npao_species)*n_pts_in_block
-    allocate(temp_block_storage(max_num_blocks))
-    
     ! loop arround grid points in the domain, and for each
     ! point, get the PAO values
 
-    do iblock = 1, domain%groups_on_node ! primary set of blocks
-       xblock=(domain%idisp_primx(iblock)+domain%nx_origin-1)*dcellx_block
-       yblock=(domain%idisp_primy(iblock)+domain%ny_origin-1)*dcelly_block
-       zblock=(domain%idisp_primz(iblock)+domain%nz_origin-1)*dcellz_block
-       if(naba_atoms_of_blocks(atomf)%no_of_part(iblock) > 0) then ! if there are naba atoms
-          iatom=0
-          do ipart=1,naba_atoms_of_blocks(atomf)%no_of_part(iblock)
-             jpart=naba_atoms_of_blocks(atomf)%list_part(ipart,iblock)
-             ind_part=DCS_parts%lab_cell(jpart)
-             do ia=1,naba_atoms_of_blocks(atomf)%no_atom_on_part(ipart,iblock)
-                iatom=iatom+1
-                ii = naba_atoms_of_blocks(atomf)%list_atom(iatom,iblock)
-                icover= DCS_parts%icover_ibeg(jpart)+ii-1
-                ig_atom= id_glob(parts%icell_beg(ind_part)+ii-1)
-
-                xatom=DCS_parts%xcover(icover)
-                yatom=DCS_parts%ycover(icover)
-                zatom=DCS_parts%zcover(icover)
-                the_species=species_glob(ig_atom)
+    ! Note: Using OpenMP in this loop requires some redesign because there is a loop
+    !       carrier dependency in next_offset_position.
+    blocks_loop: do iblock = 1, domain%groups_on_node ! primary set of blocks
+       part_in_block: if(naba_atoms_of_blocks(atomf)%no_of_part(iblock) > 0) then ! if there are naba atoms
```

Comment:
Do we need this if statement? If `naba_atoms_of_blocks(atomf)%no_of_part(iblock) == 0`, the following loop will not happen